op/main.py

```diff
@@ -33,59 +33,92 @@ def _load_metadata(charm_dir):
-        # need to be removed.
-        if not event_hook_path.is_symlink():
-            debugf(f'Hook entry at {event_hook_path} is not a symlink:'
-                   ' attempting to remove it.')
+    if event_path.exists():
+        # Non-symlink entries and the ones not pointing to the charm code file need to be removed.
+        if not event_path.is_symlink():
+            debugf(f'Path at {event_path} is not a symlink: attempting to remove it.')
             # May raise IsADirectoryError, e.g. in case it is a directory which
             # is unexpected and left to the developer or operator to handle.
-            event_hook_path.unlink()
-        elif os.readlink(event_hook_path) != 'install':
-            debugf(f'Removing entry {event_hook_path} as it does not point'
-                   ' to "install"')
-            event_hook_path.unlink()
+            event_path.unlink()
+        elif os.readlink(event_path) != charm_code_link:
+            debugf(f'Removing path {event_path} as it does not point to {charm_code_link}')
+            event_path.unlink()
         else:
             create_link = False
 
     if create_link:
-        debugf(f'Creating a new relative symlink at {event_hook_path}'
-               f' to pointing to "install" located in {charm_dir}/hooks')
-        event_hook_path.symlink_to('install')
-
-
-def _setup_hooks(charm_dir, charm):
-    """Set up hooks for supported events.
+        debugf(f'Creating a new relative symlink at {event_path} pointing to {charm_code_link}')
+        event_path.symlink_to(charm_code_link)
+
+
+def _get_charm_code_link(charm_dir):
+    def _try_link(entry):
+        if entry.is_symlink():
+            link_path = Path(os.readlink(install_entry))
+            if link_path.is_absolute():
+                raise RuntimeError(f'{charm_dir / entry} points to an absolute path, not relative: {link_path}')
+            return link_path
+        elif entry.exists():
+            raise RuntimeError(f'{charm_dir / entry} exists but is not a symlink to the charm code file')
+        else:
+            return None
+
+    install_entry = charm_dir / 'hooks' / 'install'
+    charm_code_link = _try_link(install_entry)
+    if charm_code_link is None:
+        # Kubernetes charms may not have an install entry (see: LP: #1854635) and have only a 'start' entry.
+        start_entry = charm_dir / 'hooks' / 'start'
+        charm_code_link = _try_link(start_entry)
+
+    if charm_code_link is not None:
+        charm_code_file = charm_dir / 'hooks' / charm_code_link
+        if not charm_code_file.exists():
+            raise RuntimeError(f'{charm_code_file} does not exist')
+        if not charm_code_file.is_file():
+            raise RuntimeError(f'{charm_code_file} is not a regular file')
+        return charm_code_link
+    raise RuntimeError(f'cannot determine a link to the charm code based on {charm_dir}/hooks/install and {charm_dir}/hooks/start entries')
```

Comment:
I think that all of this could be replaced with:

```python
charm_code_link = os.readlink(sys.argv[0])
```

If we're successfully executing this code, it already means that we've been started by a valid symlink, so we shouldn't really care whether that happens to be `install`, `start`, or one of the storage hooks (which are missing from this logic, currently).

Reply:
Looks like we will have some portability issues when it comes to using symlinks and `argv[0]`.

What I noticed is with `Xenial+` (python 3.5+) `argv[0]` is indeed a symlink. However, with Windows (I tested on Windows Server 2016 and python 3.7.5) it is an absolute path even if the file is a [Windows symlink](https://docs.microsoft.com/en-us/windows/win32/fileio/symbolic-links).

The logic to populate `argv[0]` is located here in cpython:
https://github.com/python/cpython/blob/3.6/Python/sysmodule.c#L2161-L2259 (3.6)
https://github.com/python/cpython/blob/3.7/Python/pathconfig.c#L280-L393 (3.7)

The Windows-specific replacement logic:
https://github.com/python/cpython/blob/3.7/Python/pathconfig.c#L352-L360
```
#if defined(MS_WINDOWS)
        /* Replace the first element in argv with the full path. */
        wchar_t *ptemp;
        if (GetFullPathNameW(argv0,
                           Py_ARRAY_LENGTH(fullpath),
                           fullpath,
                           &ptemp)) {
            argv0 = fullpath;
        }
#endif
```
Test:
```
#!/usr/bin/env python3

import sys

if __name__ == '__main__':
    print(sys.argv[0])
    import pdb
    pdb.set_trace()
```

```
C:\Users\Administrator\Downloads\test\lib\charm.py
> c:\users\administrator\downloads\test\lib\charm.py(9)<module>()
-> print(os.readlink(sys.argv[0]))
(Pdb) sys.argv[0]
'C:\\Users\\Administrator\\Downloads\\test\\lib\\charm.py
```

Reply:
Also, `pathlib` does not create valid symlinks with Linux-style paths (the target is empty) unless a `Path` object is specified or Windows-style slashes are used:

```
Python 3.7.5 (tags/v3.7.5:5c02a39a0b, Oct 14 2019, 23:09:19) [MSC v.1916 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pathlib
>>> sp = pathlib.Path('start')
>>> sp.symlink_to('../lib/charm.py')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python37-32\lib\pathlib.py", line 1337, in symlink_to
    self._accessor.symlink(target, self, target_is_directory)
FileExistsError: [WinError 183] Cannot create a file when that file already exists: '../lib/charm.py' -> 'start'
>>> sp.symlink_to('../lib/charm.py')
>>> sp.resolve()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python37-32\lib\pathlib.py", line 1151, in resolve
    s = self._flavour.resolve(self, strict=strict)
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python37-32\lib\pathlib.py", line 202, in resolve
    s = self._ext_to_normal(_getfinalpathname(s))
OSError: [WinError 123] The filename, directory name, or volume label syntax is incorrect: 'start'
```

`os.readlink` returns a Linux-style path from the Windows symlink. This is why Windows makes the target empty AFAIKS:
```
>>> import os
>>> os.readlink(sp)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: readlink() argument 1 must be str, not WindowsPath
>>> os.readlink(str(sp))
'../lib/charm.py'
```

If either `Path('../lib/charm.py')` or `symlink_to('..\lib\charm.py')` are used, a relative symlink gets created correctly (although Windows explorer shows it as absolute, using `os.readlink` or [this windows explorer extension](https://schinagl.priv.at/nt/hardlinkshellext/linkshellextension.html shows that it is relative):
```
>>> sp.unlink()
>>> spw = pathlib.Path('start')
>>> spw.symlink_to('..\lib\charm.py')
>>> spw.resolve()
WindowsPath('C:/Users/Administrator/Downloads/test/lib/charm.py')
>>> os.readlink(str(spw))
'..\\lib\\charm.py'
>>> spw.unlink()

>>> spw.symlink_to(pathlib.Path('../lib/charm.py'))
>>> os.readlink(str(spw))
'..\\lib\\charm.py'
```

Reply:
Something like this should work on Linux distributions and Windows:

```pathlib.Path(symlink_name).symlink_to(os.path.relpath(os.path.realpath(sys.argv[0]), '/tmp/test/hooks'))```

I validated that the charm store does not mangle symlinks in the zip archive (the `charm` tool that reactive uses copies a hook template for all necessary events but this is not a general charm store behavior).

The problem is the portability of symlinks extracted from zip archives because when a standard Windows tool is used to uncompress a zip charm archive all symlinks are just **regular files** that contain the paths that symlinks contained before compression.

Juju, however, uses Go's `archive/zip` and it looks like there is some logic to handle symlinks:

https://github.com/juju/charm/blob/a35e524ea871ac01a0522bd37b6267a45e233fb2/charmarchive.go#L335-L339
https://github.com/juju/utils/blob/master/zip/zip.go#L41-L112
https://github.com/juju/utils/blob/master/zip/zip.go#L160-L171 (`os.Symlink(symlinkTarget, targetPath)`)

We will need to check the actual behavior on a Windows unit but it might just work if I read it correctly.

Reply:
My simulation in the test charm shows that we cannot rely on argv[0] to determine an event name in a cross-platform manner.

test/charms/test_main/lib/charm.py:
```
if __name__ == '__main__':
    import sys
    sys.argv[0] = os.path.realpath(sys.argv[0])
    main(Charm)
```

```
[1] > /home/ubuntu/src/juju/operator/test/charms/test_main/lib/op/main.py(158)main()
-> juju_event_name = Path(sys.argv[0]).name
(Pdb++) n
[1] > /home/ubuntu/src/operator/juju/operator/test/charms/test_main/lib/op/main.py(159)main()
-> if not juju_event_name:
(Pdb++) juju_event_name
'charm.py'
```

I updated the bug https://bugs.launchpad.net/juju/+bug/1854505 with this info as we need an environment variable to reliably determine the event name.

Reply:
https://github.com/canonical/operator/pull/63/commits/e4a4df9eb3658595034bc24a96d874c724b76ab7